integrate/simpsons.go

```diff
@@ -0,0 +1,124 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import (
+	"math"
+	"sort"
+)
+
+// Simpson's method approximates the integral of a function f(x)
+// by means of subdividing the interval of integration into segments
+// and applying a method that fits a polynomial to each subinterval.
+// this implementation makes use of the following composite Simpson's method
+// to estimate \int_a^b f(x) dx:
+//
+// \sum_{i=0}^{(N/2)-1} {a_0}_i * f_{2i+1} + {a_1}_i * f_{2i+1} + {a_2}_i * f_{2i}
+//
+// where N is the number of subintervals and the number of subintervals
+// is even and {a_0}_i, {a_1}_i, and {a_2}_i are constants at index i given by:
+//
+// {a_0}_i = 2 * h^{3}_{2i+1} - h^{3}_{2i} + 3 * h_{2i} * h^{2}_{2i+1} /
+//           6 * h_{2i+1} * (h_{2i+1} + h_{2i})
+//
+// {a_1}_i = h^{3}_{2i+1} + h^{3}_{2i} +
+//           3 * h_{2i+1} * (h_{2i+1} + h_{2i}) /
+//           6 * h_{2i+1} * h_{2i}
+//
+// {a_2}_i = 2 * h^{3}_{2i} - h^{3}_{2i+1} + 3 * h_{2i+1} * h^{2}_{2i} /
+//           6 * h_{2i} * (h_{2i+1} + h_{2i})
+//
+// The formula above approximates the integral of function of f
+// if N is an even number of subintervals.
+//
+// If the number of subintervals are odd,
+// the subintervals up i=0..n-2 are given by the above
+// and the approximations over the second to last and last subintervals
+// are given by:
+//
+// {a_0}_i * f_{N} + {a_1}_i * f_{N-1} + {a_2}_i * f_{N-1}
+//
+// where the coefficients are:
+//
+// {a_0}_i = 2 * h^{2}_{N-1} + 3 * h^{2}_{N-1} * h_{N-2} /
+//           6 * (h_{N-2} + h_{N-1})
+//
+// {a_1}_i = h^{2}_{N-1} + 3 * h^{3}_{N-1} * h_{N-2} /
+//           6 * h_{N-2}
+//
+// {a_2}_i = -1 * h^{3}_{N-1} /
+//            6 * h_{N-2} * (h_{N-2} + h_{N-1})
+// More information is available at:
+// https://en.wikipedia.org/wiki/Simpson%27s_rule#Composite_Simpson's_rule_for_irregularly_spaced_data
+//
+// The (x,f) input data points must be sorted along x.
+// One can use github.com/gonum/stat.SortWeighted to do that.
+// The x and f slices must be of equal length and have length > 1.
+func Simpsons(x, f []float64) float64 {
+	validateInterval(x, f)
+
+	var integral float64
+
+	// If only three points are supplied,
+	// approximate the integral
+	// assuming the seccond coordinate points
+	// suffices as a midpoint and interpolate.
+	if len(x) == 3 {
+		return (x[2] - x[0]) * (f[0]/6 + 4/6*f[1] + 4*f[2]/6)
+	}
+
+	// Compute difference as coordinates are traversed
+	diff := []float64{}
```

Comment:
This slice is not necessary, you ever need only it's last two elements. If you don't want to compute `h0` and `h1` anew in every iteration, just hoist them out of the loop and update them.

Reply:
Removed the slice